### ISSUE TITLE
javax.imageio.IlOException: Can't create an ImageOutputStream!

### ISSUE THREAD

**Issue by AppleCarmine** [shmuelzon/home-assistant-floor-plan#91]
**Describe the bug**
When i try to render i get Failed to render floor plan: javax.imageio.IlOException: Can't create an ImageOutputStream!

**To Reproduce**
1. Go to the plugin
2. Click on Start
3. See error

![image](https://github.com/user-attachments/assets/f6b5e2af-a42e-4fef-8637-0e1302adcb68)

**Comment by shmuelzon** [comment]
Hey, it looks like it failed to save a rendered image to disk. The only reason I can think of is write permissions to the directory but I would have assumed it would have also failed creating the `render` and `floorplan` directories. Have those been created? Have those been created with proper permissions to write in them?

**Comment by AppleCarmine** [comment]
> Ehi, sembra che non sia riuscito a salvare un'immagine renderizzata su disco. L'unica ragione che mi viene in mente sono i permessi di scrittura sulla directory, ma avrei pensato che non sarebbe riuscito a creare le directory `render`e `floorplan`. Sono state create? Sono state create con i permessi di scrittura appropriati?

On Mac, I granted full disk access to Sweet Home 3D: nothing. I also tried on Windows, but it doesn’t work.

**Comment by shmuelzon** [comment]
Maybe I'm misreading the error message...
Would you mind sharing, even privately, your SH3D project file so I can see if it reproduces on my setup as well?
BTW - which version of SH3D are you using?

**Comment by AppleCarmine** [comment]
> Maybe I'm misreading the error message... Would you mind sharing, even privately, your SH3D project file so I can see if it reproduces on my setup as well? BTW - which version of SH3D are you using?

Of course! I’ll send you the file privately. I tried both version 7.5.5 and version 7.5.1. Where can I send the file?

**Comment by shmuelzon** [comment]
You can email it to me directly. Same user as here at gmail.com

**Comment by shmuelzon** [comment]
Well, it rendered just fine for me on Windows with v0.7.3 of this plugin and SH3D v7.5 (don't know which minor version since I don't see it anywhere but it's not asking me to update, so 🤷)
![image](https://github.com/user-attachments/assets/9c6fab70-ee2c-4c1e-a768-a0ffae2fac46)
![image](https://github.com/user-attachments/assets/3fcda3cb-2f80-4e80-9255-6b483c8d5f56)
![image](https://github.com/user-attachments/assets/b29b3f40-f447-42e4-af0a-556e8e65cc0b)

I'm not sure what else do to. Do you see the same error when using Windows or is it something else?

**Comment by AppleCarmine** [comment]
> Well, it rendered just fine for me on Windows with v0.7.3 of this plugin and SH3D v7.5 (don't know which minor version since I don't see it anywhere but it's not asking me to update, so 🤷) ![image](https://private-user-images.githubusercontent.com/2786045/387710450-9c6fab70-ee2c-4c1e-a768-a0ffae2fac46.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMzc0NTcsIm5iZiI6MTczMjAzNzE1NywicGF0aCI6Ii8yNzg2MDQ1LzM4NzcxMDQ1MC05YzZmYWI3MC1lZTJjLTRjMWUtYTc2OC1hMGZmYWUyZmFjNDYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTlUMTcyNTU3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MGVlYzE5NjEzNGNkODcwY2JiZTk2ZjNlZjBhNzM5MzE0YjEzNGI4MDk1NTJhMThhMzI5Mjk1NWQ0NzMzNGIyOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.u1cVxxKs4xS-M2ea5s9-hbGVUs5oSVRSshgP_jJvuBA) ![image](https://private-user-images.githubusercontent.com/2786045/387710700-3fcda3cb-2f80-4e80-9255-6b483c8d5f56.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMzc0NTcsIm5iZiI6MTczMjAzNzE1NywicGF0aCI6Ii8yNzg2MDQ1LzM4NzcxMDcwMC0zZmNkYTNjYi0yZjgwLTRlODAtOTI1NS02YjQ4M2M4ZDVmNTYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTlUMTcyNTU3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmY0ZTkzMmZmZDY1NTA5OWM0ZGVhYTkwZGU4Zjc0YzFjOGVjMTk3MjU0YTFkZTBjN2ZkOTMwMGY4MDJiZDU5NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Q_kY7XAWJTmqCkaGz7HmFsYTP453QoDQNzchR4cvfgc) ![image](https://private-user-images.githubusercontent.com/2786045/387710785-b29b3f40-f447-42e4-af0a-556e8e65cc0b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMzc0NTcsIm5iZiI6MTczMjAzNzE1NywicGF0aCI6Ii8yNzg2MDQ1LzM4NzcxMDc4NS1iMjliM2Y0MC1mNDQ3LTQyZTQtYWYwYS01NTZlOGU2NWNjMGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTlUMTcyNTU3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDZmMzVlZjk0YjQ3YzBmZDUwMWYzNzNhMjJlNTIzOGUyYjI2MWU4MGQ0YTEwZGY1NjFkODI1ZWM5OTRiYWIzMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.7UR13mcm7hr88JR4GVRUmFWl4io8XTRaGdnsY4F_kH8)
> 
> I'm not sure what else do to. Do you see the same error when using Windows or is it something else?

Yes, same error on Windows, tried with Java 8 and Java 17. Idk what should i do, i also tried with different pc.

**Comment by AppleCarmine** [comment]
> How do you mean different Java versions? The Windows installation already comes bundled with Java. Are you using the `jar` directly or something?
> 
> I can build the plugin to give a full trace so we'll be sure where the error comes from, maybe that can lead to something? BTW - are the `render` and `floorplan` directories created in the output directory you set?
> 
> Edit: built a v0.7.3 with a full stack trace on error: [HomeAssistantFloorPlanPlugin-v0.7.3.zip](https://github.com/user-attachments/files/17820001/HomeAssistantFloorPlanPlugin-v0.7.3.zip)

No, the render and floorplan folders are not created. By Java, I mean the version of Java installed on the PC.

**Comment by AppleCarmine** [comment]
> How do you mean different Java versions? The Windows installation already comes bundled with Java. Are you using the `jar` directly or something?
> 
> I can build the plugin to give a full trace so we'll be sure where the error comes from, maybe that can lead to something? BTW - are the `render` and `floorplan` directories created in the output directory you set?
> 
> Edit: built a v0.7.3 with a full stack trace on error: [HomeAssistantFloorPlanPlugin-v0.7.3.zip](https://github.com/user-attachments/files/17820001/HomeAssistantFloorPlanPlugin-v0.7.3.zip)

![image](https://github.com/user-attachments/assets/3292c098-859a-46bd-a23d-678acb6458fd)

**Comment by shmuelzon** [comment]
Hey, so, first of all, I fixed the plugin to let one know that it couldn't create the output directories since there's no reason to wait for a render to complete when that happens. Can you please try it too? It might give a more specific error in your case. The image writing one could just have been because the directory doesn't exist 🤷‍♂️.

As for the actual problem, have you tried setting the output directory to a globally writeable one, like `/tmp` or something? Does it still return the same error?

New version: [HomeAssistantFloorPlanPlugin-v0.7.3.zip](https://github.com/user-attachments/files/17825630/HomeAssistantFloorPlanPlugin-v0.7.3.zip)

**Comment by AppleCarmine** [comment]
> Ciao, quindi, prima di tutto, ho sistemato il plugin per far sapere a uno che non poteva creare le directory di output poiché non c'è motivo di aspettare che un rendering venga completato quando ciò accade. Puoi provare anche tu? Potrebbe dare un errore più specifico nel tuo caso. L'immagine che scrive uno potrebbe essere semplicemente perché la directory non esiste 🤷‍♂️.
> 
> Per quanto riguarda il problema effettivo, hai provato a impostare la directory di output su una scrivibile globalmente, tipo `/tmp`o qualcosa del genere? Restituisce ancora lo stesso errore?
> 
> Nuova versione: [HomeAssistantFloorPlanPlugin-v0.7.3.zip](https://github.com/user-attachments/files/17825630/HomeAssistantFloorPlanPlugin-v0.7.3.zip)

![image](https://github.com/user-attachments/assets/73cc04f6-ba3e-4e78-a726-286e34abc210)
sorry for replying only now, but i didn't see your reply. Here's the error with the version you just gave me:

**Comment by AppleCarmine** [comment]
![image](https://github.com/user-attachments/assets/745b2d4c-132c-45a0-98e4-3fe1746291a6)
I tried to manually create the renders and floorplan folders, this is the error I then get

**Comment by shmuelzon** [comment]
I'm at a loss here... It looks like some Apple sandboxing thing but you say you've already enabled full disk access. May the SH3D launcher just runs a Java process and then that means we need to give that process full disk access?

**Comment by AppleCarmine** [comment]
> Sono indeciso... Sembra una specie di sandboxing Apple, ma dici di aver già abilitato l'accesso completo al disco. Forse il launcher SH3D esegue solo un processo Java e quindi ciò significa che dobbiamo dare a quel processo l'accesso completo al disco?

How could I check? The strangest thing is that it doesn’t even work properly from Windows.

**Comment by assafi-liveu** [comment]
Unfortunately, I'm not one for creating videos. There are many tutorials for SH3D in general and the README file for this project should include everything you need to generate a Home Assistant floor plan from your model.

**Comment by shmuelzon** [comment]
Unfortunately, I'm not one for creating videos. There are many tutorials for SH3D in general and the README file for this project should include everything you need to generate a Home Assistant floor plan from your model.

**Comment by shmuelzon** [comment]
Hey, I pushed #95 which will show the error message in case it can't create the directories which should, at least, let someone know right away when there's a problem instead of waiting for the base image render to complete.

As for the macOS specific issue, I don't have access to a Mac so I can't debug this on my own. I'll leave this ticket open, though, in the hopes that someone else could help out.

**Comment by resnbl** [comment]
Initially I had the same issue here on a new Mac Mini M4 running Sequoia 15.1.1 with v0.7.4: quick fail until I manually created the directories; then it apparently failed when it tried to write the first render. Note I also had to manually type in my path in the dialog box: the browse button opened a dialog in which nothing could be selected and the main /User directory could not even be opened.
Then I remembered why I "nuked from space" my entire Java environment when I retired a number of years ago and started goofing:
First stab was to replace the base path's "/" directory separator characters conventionally used on Mac/Unix machines with the (dreadful) Windows variety: "\\".

Damn if it didn't work on the first try!

![Screenshot 2024-12-05 at 6 54 29 PM](https://github.com/user-attachments/assets/9230e093-0280-41ad-9b88-bf447bcf1df8)

Never thought I would use backward slashes on a Mac...

Let me know if you would like me to test any other modifications (like localizing the Render Time for we who freak at DD/MM/YYYY).

**Comment by shmuelzon** [comment]
@resnbl  thanks for that!
It's odd that you say it seemed to run fine since it first writes the images to the renders directory and then reads them back for the floorplan directory so they should have existed somewhere.

Have you also tried enabling "full disk access" to SH3D (or the Java binaries that are bundled with it) in the OS security setting and see if that makes a difference with the browse button and/or actual renders?

**Comment by resnbl** [comment]
I made some progress: turns out the files WERE created, but not where I would expect them. I found them in `/Users/bob/Library/Containers/com.eteks.sweethome3d.SweetHome3D/Data/\\User\\bob\\floorp/` (note the backslashes are actually part of the directory name!).
This definitely has something to do with the fact that SH3D must be a "sandboxed" app in order to be sold in the Apple Mac Store. The `/Users.../Data` portion of the path is the area allocated for each sandboxed app. I suspect your Java file system methods are operating directly on the main user directory, but the SH3D file I/O is occurring in the sandbox directory.
I'll try some more later to see if I can suss out what needs to be entered where in order to get this to work...

**Comment by shmuelzon** [comment]
Maybe on macOS the directory needs to be relative (without the leading forward slash) and send it works from the sandboxed directory. With backslashes, as you pointed out, it just used them as "regular" characters that are part of the name. If you just enter `./` or something as the destination directory, would it create the files in `/Users/bob/Library/Containers/com.eteks.sweethome3d.SweetHome3D/Data/`?

BTW - How does SH3D work, in general. on macOS? When you load/save a project, can you do that from any directory you want, outside of the sandboxed environment. If so, I need to understand what the difference is and why that's allowed but not the way I use it.

**Comment by resnbl** [comment]
Yes, the Mac OS version can read/write .sh3d files in normal directories (once you give the app proper folder access).

I was able to generate files with a simple folder name ("floorp"), however I need to retest to determine the exact prerequisites. (Must /Users/bob/floorp be defined? How about the sub-dirs?...)

**Comment by shmuelzon** [comment]
Thank you for putting time into this! Even if we don't find a solution to make this work in any directory, at least we can add something to the FAQ as a workaround to make it usable.

**Comment by shmuelzon** [comment]
From what I understand, if an application uses the file dialog for loading/saving a specific file, it works fine for sandboxing as the user specified exactly which file he wants to access. I guess it's not the same when when selecting a directory, though, as that's too permissive so that can only be done in the sandbox directory.
I'm still not sure why it doesn't work when the application is provided full disk access, though. That should have allowed it to access the entire file system

**Comment by resnbl** [comment]
Steps for the best I can do:
1. Create a folder in home directory (just for existence check): `/User/bob/floorp`
2. Open extension in SH3D and manually change the output directory to just the folder name: `floorp`
3. Start the render.
4. The files will be output to: `/Users/bob/Library/Containers/com.eteks.sweethome3d.SweetHome3D/Data/floorp/`
5. Note that this directory cannot be easily accessed in Finder (Mac's "File Explorer"). To show it, use the Finder "Go To..." menu and enter the full path above. Alternatively, you can `cd` in the Terminal (console) app.

The issue here is that sandboxed Mac apps cannot arbitrarily write outside the app's data folder unless the folder is selected in a **system** file dialog which is how SH3D can access files in arbitrary locations. (Note the Java FX/Swing dialog shown here doesn't cut it as it never enables the **Select** button).
![browse](https://github.com/user-attachments/assets/53b47d8c-0180-458b-b2d6-340e15dc5d90)

My only suggestion is to see if the dialog SH3D uses can be modified to choose a directory. This may require an additional property key to be set to operate correctly.

**Comment by MrNoooooooob** [comment]
> Steps for the best I can do:
> 
> 1. Create a folder in home directory (just for existence check): `/User/bob/floorp`
> 2. Open extension in SH3D and manually change the output directory to just the folder name: `floorp`
> 3. Start the render.
> 4. The files will be output to: `/Users/bob/Library/Containers/com.eteks.sweethome3d.SweetHome3D/Data/floorp/`
> 5. Note that this directory cannot be easily accessed in Finder (Mac's "File Explorer"). To show it, use the Finder "Go To..." menu and enter the full path above. Alternatively, you can `cd` in the Terminal (console) app.
> 
> The issue here is that sandboxed Mac apps cannot arbitrarily write outside the app's data folder unless the folder is selected in a **system** file dialog which is how SH3D can access files in arbitrary locations. (Note the Java FX/Swing dialog shown here doesn't cut it as it never enables the **Select** button). ![browse](https://private-user-images.githubusercontent.com/33208090/393462015-53b47d8c-0180-458b-b2d6-340e15dc5d90.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUwMzM1NTEsIm5iZiI6MTczNTAzMzI1MSwicGF0aCI6Ii8zMzIwODA5MC8zOTM0NjIwMTUtNTNiNDdkOGMtMDE4MC00NThiLWIyZDYtMzQwZTE1ZGM1ZDkwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI0VDA5NDA1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhZDAwZWZmZGY1OTgxZGFhODVkNTZmOWE0N2ZlZDJkMDgwYjM3ODM1ZWNmNTc5YmExNzg2MTZlOTUxMTgyYjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.HaVkhHxq8wwqYEKxUlDatq2ElgZxM9MJfo2vg-P3-PQ)
> 
> My only suggestion is to see if the dialog SH3D uses can be modified to choose a directory. This may require an additional property key to be set to operate correctly.

The commands for the invisible files should also be mentioned.


Open `defaults write com.apple.Finder AppleShowAllFiles true`
`killall Finder`
Closed
`defaults write com.apple.Finder AppleShowAllFiles false`
`killall Finder`

**Comment by resnbl** [comment]
I needed to update my floor plan after moving some things around and ran into another "weirdness". The output directory under `/Users/xxx/Library/Containers/com.eteks.sweethome3d.SweetHome3D/Data` is no longer visible in the Finder. Instead it is now under `/Users/xxx/Library/Containers/Sweet Home 3D/Data`. The funny thing is that this new directory is not visible from the command line: only the prior directory shows up. I don't know is this is something that happened when I updated SH3D in March, or if this is the result of an OS change.

Further spelunking shows there are other non-Apple apps that have this type of "alias" feature going on. I wonder if I just missed it the previous time around...

**Comment by shmuelzon** [comment]
@resnbl thanks for that. Unfortunately, as I don't own a Mac I cannot test this or even try to overcome it somehow. The best I can currently offer is adding a section in the F.A.Q for this MacOS sandboxing. I'll try to write up something based on all the information here in the near future.

**Comment by Robiv8** [comment]
hallo, i have the same problem on a
iMac 5K, 27" 2017 4,2 GHz Quad-Core Intel Core i7

Sweet Home 3D ® Version 7.6.3 Mac App Store (1547)
OpenJDK Version 15.0.10 - 64bit - 0,2 / 4 GB gestellt von [Azul Systems, Inc.](http://www.azul.com/) Java 3D Version 1.6.2a - JOGL 2.5.0 gestellt von [JogAmp Community](http://jogamp.org/) SVG-Export zur Verfügung gestellt von [FreeHEP](http://java.freehep.org/) SVG-Path compilation zur Verfügung gestellt von [Batik](http://xmlgraphics.apache.org/batik/) Math expressions support gestellt von [JeksParser](http://www.eteks.com/jeks/en/) Photo Rendering zur Verfügung gestellt von [SunFlow](http://sunflow.sourceforge.net/) und [YafaRay](https://github.com/YafaRay)

any tip how to fix this

<img width="593" height="791" alt="Image" src="https://github.com/user-attachments/assets/58277eba-48c5-4599-a092-c16d000a504c" />

Thanks
robi

**Comment by resnbl** [comment]
See the list of process steps above and follow them precisely. I noticed your output directory is not set to exactly 1 folder name (I used "floorp" but you can use another name - save yourself potential headaches and don't include spaces or other special characters). This program cannot access the Desktop.
